netplugin/nameserver/nameserver.go

```diff
@@ -0,0 +1,710 @@
+		tenant, eps.EndpointID, eps.EndpointGroupKey, eps.EPCommonName,
+		eps.IPAddress, eps.IPv6Address)
+	tenMap := ens.getBucket(tenant)
+
+	tenMap.RWMutex.Lock()
+	defer tenMap.RWMutex.Unlock()
+
+	tenantTables, ok := tenMap.tenantTables[tenant]
+	if !ok {
+		tenantTables = new(dnsTables)
+		tenMap.tenantTables[tenant] = tenantTables
+	}
+
+	//update endpoint
+	if tenantTables.endpointTbl == nil {
+		tenantTables.endpointTbl = make(map[string]nameRecord)
+	}
+
+	epEntry := nameRecord{
+		v4Record: net.ParseIP(eps.IPAddress),
+		v6Record: net.ParseIP(eps.IPv6Address),
+	}
+
+	tenantTables.endpointTbl[eps.EndpointID] = epEntry
+
+	//update name
+	if len(eps.EPCommonName) > 0 {
+		containerName := eps.EPCommonName
+		if eps.EPCommonName[:1] == "/" {
+			containerName = eps.EPCommonName[1:]
+		}
+
+		if tenantTables.nameTbl == nil {
+			tenantTables.nameTbl = make(map[string]map[string]bool)
+		}
+		nl, ok := tenantTables.nameTbl[containerName]
+		if !ok {
+			nl = make(map[string]bool)
+			tenantTables.nameTbl[containerName] = nl
+		}
+		nl[eps.EndpointID] = true
+	}
+
+	//update epg
+	if len(eps.EndpointGroupKey) > 0 {
+		if tenantTables.epgTbl == nil {
+			tenantTables.epgTbl = make(map[string]map[string]bool)
+		}
+		epgName := ens.getEpgName(eps.EndpointGroupKey)
+		epgList, ok := tenantTables.epgTbl[epgName]
+		if !ok {
+			epgList = make(map[string]bool)
+			tenantTables.epgTbl[epgName] = epgList
+		}
+		epgList[eps.EndpointID] = true
+	}
+
+}
+
+func (ens *NetpluginNameServer) delEndpoint(s core.State) {
+	eps, ok := s.(*mastercfg.CfgEndpointState)
+
+	if !ok {
+		ens.incTenantErrStats("", "endPointDel")
+		return
+	}
+
+	tenant := ens.getTenantName(eps.NetID)
+	dnsLog.Infof("[tenant: %s]delete endpoint: %s, epg: %s name:%s ipv4:%s ipv6:%s", tenant,
+		eps.EndpointID, eps.EndpointGroupKey, eps.EPCommonName,
+		eps.IPAddress, eps.IPv6Address)
+	tenMap := ens.getBucket(tenant)
+	tenMap.RWMutex.Lock()
+	defer tenMap.RWMutex.Unlock()
+	if tenantTables, ok := tenMap.tenantTables[tenant]; ok {
+		if tenantTables.epgTbl != nil && len(eps.EndpointGroupKey) > 0 {
+			epgName := ens.getEpgName(eps.EndpointGroupKey)
+			epgList, ok := tenantTables.epgTbl[epgName]
+			if ok {
+				delete(epgList, eps.EndpointID)
+				if len(epgList) <= 0 {
+					delete(tenantTables.epgTbl, epgName)
+				}
+			}
+
+		}
+		if tenantTables.nameTbl != nil && len(eps.EPCommonName) > 0 {
+			containerName := eps.EPCommonName
+			if eps.EPCommonName[:1] == "/" {
+				containerName = eps.EPCommonName[1:]
+			}
+			nameList, ok := tenantTables.nameTbl[containerName]
+			if ok {
+				delete(nameList, eps.EndpointID)
+				if len(nameList) <= 0 {
+					delete(tenantTables.nameTbl, containerName)
+				}
+			}
+		}
+		if tenantTables.endpointTbl != nil {
+			delete(tenantTables.endpointTbl, eps.EndpointID)
+		}
+	}
+}
+
+func (ens *NetpluginNameServer) incTenantStats(tenant string, name string) {
+	ens.stats.RWMutex.Lock()
```

Comment:
RWMutex doesn't need to be called explicitly here. You can call Lock/Unlock directly on stats as you have done in the next step. Please do the same for tenant map